core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -257,6 +277,100 @@ public Response call() throws S3Exception {
   }
 
   /**
+   * @summary initiates or completes a multipart upload based on query parameters
+   * @param bucket the bucket name
+   * @param object the object name
+   * @param uploads the query parameter specifying that this request is to initiate a multipart
+   *                upload instead of uploading an object through HTTP multipart forms
+   * @param uploadId the ID of the multipart upload to be completed
+   * @return the response object
+   */
+  @POST
+  @Path(OBJECT_PARAM)
+  @ReturnType("alluxio.proxy.s3.InitiateMultipartUploadResult,"
+      + "alluxio.proxy.s3.CompleteMultipartUploadResult")
+  // TODO(cc): investigate on how to specify multiple return types, and how to decouple the REST
+  // endpoints where the only difference is the query parameter.
+  public Response initiateOrCompleteMultipartUpload(@PathParam("bucket") final String bucket,
+      @PathParam("object") final String object, @QueryParam("uploads") final String uploads,
+      @QueryParam("uploadId") final Long uploadId) {
+    Preconditions.checkArgument(uploads != null || uploadId != null,
+        "parameter 'uploads' or 'uploadId' should exist");
+    if (uploads != null) {
+      return initiateMultipartUpload(bucket, object);
+    } else {
+      return completeMultipartUpload(bucket, object, uploadId);
+    }
+  }
+
+  private Response initiateMultipartUpload(final String bucket, final String object) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<InitiateMultipartUploadResult>() {
+      @Override
+      public InitiateMultipartUploadResult call() throws S3Exception {
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        String objectPath = bucketPath + AlluxioURI.SEPARATOR + object;
+        AlluxioURI multipartTemporaryDir =
+            new AlluxioURI(S3RestUtils.getMultipartTemporaryDirForObject(bucketPath, object));
+
+        try {
+          mFileSystem.createDirectory(multipartTemporaryDir);
+          long uploadId = mFileSystem.getStatus(multipartTemporaryDir).getFileId();
```

Comment:
leave a comment about using fileId of temp dir as uploadId?

Reply:
done